e2e/test/MessageReceiveE2ETests.cs

```diff
@@ -234,6 +295,53 @@ public static async Task VerifyReceivedC2DMessageAsync(Client.TransportType tran
             Assert.IsTrue(received, $"No message received for device {deviceId} with payload={payload} in {FaultInjection.RecoveryTimeMilliseconds}.");
         }
 
+        public static async Task VerifyReceivedC2DMessageWithCancellationTokenAsync(Client.TransportType transport, DeviceClient dc, string deviceId, string payload, string p1Value)
+        {
+            Stopwatch sw = new Stopwatch();
+            bool received = false;
+
+            sw.Start();
+
+            Client.Message receivedMessage;
+
+            while (!received && sw.ElapsedMilliseconds < FaultInjection.RecoveryTimeMilliseconds)
+            {
+                _log.WriteLine($"Receiving messages for device {deviceId}.");
+
+                
+                receivedMessage = await dc.ReceiveAsync(new CancellationTokenSource(TIMESPAN_ONE_MINUTE).Token).ConfigureAwait(false);
+                
+                if (receivedMessage == null)
+                {
+                    Assert.Fail($"No message is received for device {deviceId} in {TIMESPAN_ONE_MINUTE}.");
+                }
+
+                try
+                {
+                    // always complete message
+                    await dc.CompleteAsync(receivedMessage).ConfigureAwait(false);
+                }
+                catch (Exception)
+                {
+                    // ignore exception from CompleteAsync
+                }
+
+                string messageData = Encoding.ASCII.GetString(receivedMessage.GetBytes());
```

Comment:
duplicate code, can be shared with other verify tasks

Reply:
Sure. It could be extracted. But it is coupling with `while` loop. It could be extracted as `bool VerifyMessage(Client.Message message, string payload, string payloadKey, string payloadValue)`. It will return `false` if `payload` doesn't match.